epf/BufferCache.cpp

```diff
@@ -21,9 +21,11 @@ namespace epf
 {
 
 // If we have a buffer in the cache, return it. Otherwise create a new one and return that.
-DataVecPtr BufferCache::fetch()
+// If nonblock is true and there are no available buffers, return null.
+DataVecPtr BufferCache::fetch(std::unique_lock<std::mutex>& lock, bool nonblock)
 {
-    std::unique_lock<std::mutex> lock(m_mutex);
+    if (nonblock && m_buffers.empty() && m_count >= MaxBuffers)
```

Comment:
Should this `buffers.empty()` check be locked?

Reply:
Actually, no. Or to put it another way, it already is.

The lock that was in the BufferCache has been removed. The lock that's used is now the one that's in the Writer, that owns the buffer cache. `Writer::fetchBuffer()` acquires the lock before calling `BufferCache::fetch()`. It passes the lock along so that it can be UN-locked if we need to wait on a CV.